### PR TITLE
Noise

### PR DESCRIPTION
**Pull request by mattradz**
Basic noise, noise_test.py, changed devices

### CODE REVIEW

File: qnetdes/noise.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    program += noisy_X_definition
+    
+    #apply
+    if np.random.rand() > prob:
+        program += flip_NOISY_X(qubit)
+    else:
+        program += flip_NOISY_I(qubit)
+    
+
+def phase_flip(program, qubit, prob: float):
```
Comment (zacespinosa) [inline qnetdes/noise.py]: I think `DefGate` needs to be done globally, once, for each function

File: qnetdes/noise.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+def normal_unitary_rotation(program, qubit, prob:float):
+    '''
+    Apply X and Z rotation with probability
+
+    :param Program program: program to apply noise to
+    :param Integer qubit: qubit to apply noise to 
+    :param Float prob: probability of apply noise 
+    '''
+    if np.random.rand() > prob:
+        x_angle, z_angle = np.random.normal(0,self.variance,2)
```
Comment (zacespinosa) [inline qnetdes/noise.py]: self.variance is not defined here

File: qnetdes/devices.py [hunk 3]
```diff
@@ -55,44 +53,42 @@ def __init__(self, pulse_length=10 * 10 ** -12, expected_photons=1.0, rotation_p
...
         for qubit in qubits:
             if qubit is not None and self.apply_error:
                 numPhotons = np.random.poisson(lam=self.photon_expectation)
-                print("Produced ", numPhotons, " photons")
+                self.trials += 1
+                if numPhotons == 1: self.success += 1
+                
+                
+                '''
+                #Rotation noise
```
Comment (zacespinosa) [inline qnetdes/devices.py]: This should use `normal_unitary_rotation`

File: qnetdes/noise.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+def bit_flip(program, qubit, prob: float):
+    '''
+    Apply a bit flip with probability
+
+    :param Program program: program to apply noise to
+    :param Integer qubit: qubit to apply noise to 
+    :param Float prob: probability of apply noise 
+    '''
+    #define the gates
+    noisy_I = np.asarray([[1, 0], [0, 1]])
```
Comment (zacespinosa) [inline qnetdes/noise.py]: Why not just use the pyquil version of all of these gates for all of these functions?

File: qnetdes/devices.py [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 from pyquil import Program
 from pyquil.gates import *
-from qnetdes import *
+from qnetdes import noise
```
[the remark this answers is not in the document]
Comment (mattradz) [reply qnetdes/devices.py]: ok i did this, but now moved apply_error as a parameter to the front of init for each device

File: qnetdes/noise.py (hunk 4 above)
Comment (mattradz) [reply qnetdes/noise.py]: I was going to possibly change the values such that is a "noisy" gate, namely use 1.0 which makes the gate a float and not a perfect I,X,Y,Z flip

Comment (zacespinosa) [reply qnetdes/noise.py]: dope, sounds good

File: qnetdes/noise.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+def measure(program, qubit, prob: float, name="ro"):
+    '''
+    Measure the qubit with probability
+
+    :param Program program: program to apply noise to
+    :param Integer qubit: qubit to apply noise to 
+    :param Float prob: probability of apply noise 
+    '''
+    if np.random.rand()> prob:
+        global ro_declared
```
[the remark this answers is not in the document]
Comment (mattradz) [reply qnetdes/noise.py]: i don't think this is valid syntax.

File: qnetdes/noise.py [hunk 7]
```diff
@@ -117,24 +107,25 @@ def measure(program, qubit, prob: float, name="ro"):
...
+                    pass
+
             ro = program.declare("ro", 'BIT', 1)
             ro_declared = True
             program += MEASURE(qubit, ro)
 
-
-
-def normal_unitary_rotation(program, qubit, prob:float):
+        elif ro_declared:
```
Comment (zacespinosa) [inline qnetdes/noise.py]: nit: could just be else

File: qnetdes/noise.py [hunk 8]
```diff
@@ -143,7 +134,7 @@ def normal_unitary_rotation(program, qubit, prob:float):
     :param Float prob: probability of apply noise
```
Comment (zacespinosa) [inline qnetdes/noise.py]: add :param Float variance

**Pull request merged**